src/ucp/core/ucp_mm.c

```diff
@@ -1386,20 +1398,37 @@ ucp_mem_rcache_mem_reg_cb(void *ctx, ucs_rcache_t *rcache, void *arg,
                           ucs_rcache_region_t *rregion,
                           uint16_t rcache_mem_reg_flags)
 {
-    ucp_context_h context = (ucp_context_h)ctx;
-    ucp_mem_h memh        = ucs_derived_of(rregion, ucp_mem_t);
-    ucp_mem_attr_t *attr  = arg;
-    const void *address   = (void*)memh->super.super.start;
-    size_t length         = memh->super.super.end - memh->super.super.start;
+    ucp_context_h context             = (ucp_context_h)ctx;
+    ucp_mem_rcache_reg_ctx_t *reg_ctx = arg;
+    ucp_mem_h memh                    = ucs_derived_of(rregion, ucp_mem_t);
     ucp_memory_info_t info;
+    ucs_status_t status;
+    unsigned uct_flags;
 
-    ucp_memory_detect(context, address, length, &info);
+    ucp_memory_detect(context, (void*)memh->super.super.start,
+                      memh->super.super.end - memh->super.super.start, &info);
     memh->context        = context;
     memh->alloc_md_index = UCP_NULL_RESOURCE;
     memh->alloc_method   = UCT_ALLOC_METHOD_LAST;
-    memh->mem_type       = attr->mem_type;
+    memh->mem_type       = reg_ctx->mem_type;
     memh->sys_dev        = info.sys_dev;
 
+    uct_flags = reg_ctx->uct_flags;
+    if (rcache_mem_reg_flags & UCS_RCACHE_MEM_REG_HIDE_ERRORS) {
+        uct_flags |= UCT_MD_MEM_FLAG_HIDE_ERRORS;
+    }
+
+    status = ucp_memh_register(context, memh, reg_ctx->reg_md_map, uct_flags,
+                               reg_ctx->alloc_name);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    if (reg_ctx->reg_md_map != memh->md_map) {
```

Comment:
if UCT_MD_MEM_FLAG_HIDE_ERRORS we return UCS_OK even if not all registered

Reply:
but when rcache was in UCT, we failed if it was not registered. we have to fail here to indicate that merge should be aborted..
anyway, reworked the function to return failure in such case in the next commit